flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java

```diff
@@ -0,0 +1,269 @@
+                        new ExecutorThreadFactory(
+                                "periodic-materialization-scheduler-" + subtaskName));
+    }
+
+    public void start() {
+        if (!started) {
+
+            started = true;
+
+            LOG.info(
+                    "Task {} starts periodic materialization, scheduling the next one in {} seconds",
+                    subtaskName,
+                    periodicMaterializeDelay / 1000);
+
+            scheduleNextMaterialization();
+        }
+    }
+
+    private void triggerMaterialization() {
+        mailboxExecutor.execute(
+                () -> {
+                    Optional<MaterializationRunnable> materializationRunnableOptional =
+                            keyedStateBackend.initMaterialization();
+
+                    if (materializationRunnableOptional.isPresent()) {
+                        MaterializationRunnable runnable = materializationRunnableOptional.get();
+                        asyncOperationsThreadPool.execute(
+                                () ->
+                                        asyncMaterializationPhase(
+                                                runnable.getMaterializationRunnable(),
+                                                runnable.getMaterializedTo()));
+                    } else {
+                        scheduleNextMaterialization();
+
+                        LOG.info(
+                                "Task {} has no state updates since last materialization, "
+                                        + "skip this one and schedule the next one in {} seconds",
+                                subtaskName,
+                                periodicMaterializeDelay / 1000);
+                    }
+                },
+                "materialization");
+    }
+
+    private void asyncMaterializationPhase(
+            RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture,
+            SequenceNumber upTo) {
+
+        SnapshotResult<KeyedStateHandle> materializedSnapshot =
+                uploadSnapshot(materializedRunnableFuture);
+
+        // if succeed, update state and finish up
+        if (materializedSnapshot != null) {
+
+            numberOfConsecutiveFailures.set(0);
+
+            final SnapshotResult<KeyedStateHandle> copyMaterializedSnapshot = materializedSnapshot;
+
+            mailboxExecutor.execute(
+                    () ->
+                            keyedStateBackend.updateChangelogSnapshotState(
+                                    copyMaterializedSnapshot,
+                                    upTo),
+                    "Task {} update materializedSnapshot up to changelog sequence number: {}",
+                    subtaskName,
+                    upTo);
+        }
+
+        LOG.info(
+                "Task {} schedules the next materialization in {} seconds.",
+                subtaskName,
+                periodicMaterializeDelay / 1000);
+
+        scheduleNextMaterialization();
```

Comment:
If the retry limit is exceeded, the task will be failed and trigger failover

Do you mean it is possible that before failover is triggered, another materialization is triggered?

Reply:
What's the problem if `scheduleNextMaterialization` is called?

Reply:
> Do you mean it is possible that before failover is triggered, another materialization is triggered?
> And what's the problem if scheduleNextMaterialization is called?

Yes. One problem is that it likely will fail (because some of the executors is shut/ting down),  and potentially prevent failover. If not, it's just unnecessary upload.

Besides that, the code would be much cleaner IMO.